src/primitive/polyhedron.rs

```diff
@@ -0,0 +1,713 @@
+where
+    S: BaseFloat,
+{
+    type Aabb = Aabb3<S>;
+
+    fn get_bound(&self) -> Aabb3<S> {
+        self.bound.clone()
+    }
+}
+
+/// TODO: better algorithm for finding faces to intersect with?
+impl<S> Discrete<Ray3<S>> for ConvexPolyhedron<S>
+where
+    S: BaseFloat,
+{
+    /// Ray must be in object space
+    fn intersects(&self, ray: &Ray3<S>) -> bool {
+        let mut face_index = 0;
+        let mut checked = BitSet::with_capacity(self.faces.len());
+        while face_index < self.faces.len() {
+            checked.insert(face_index);
+            let f = &self.faces[face_index];
+            match intersect_ray_face(ray, &self, &f) {
+                Some((u, v, w)) => {
+                    if in_range(u) && in_range(v) && in_range(w) {
+                        return true;
+                    } else {
+                        face_index =
+                            next_face_classify(self, face_index, Some((u, v, w)), &mut checked);
+                    }
+                }
+                _ => {
+                    face_index = next_face_classify(self, face_index, None, &mut checked);
+                }
+            }
+        }
+        false
+    }
+}
+
+// TODO: better algorithm for finding faces to intersect with?
+impl<S> Continuous<Ray3<S>> for ConvexPolyhedron<S>
+where
+    S: BaseFloat,
+{
+    type Result = Point3<S>;
+
+    /// Ray must be in object space
+    fn intersection(&self, ray: &Ray3<S>) -> Option<Point3<S>> {
+        let mut face_index = 0;
+        let mut checked = BitSet::with_capacity(self.faces.len());
+        while face_index < self.faces.len() {
+            checked.insert(face_index);
+            let f = &self.faces[face_index];
+            match intersect_ray_face(ray, &self, &f) {
+                Some((u, v, w)) => {
+                    println!("face v: {:?}, {:?}, {:?}", u, v, w);
+                    if in_range(u) && in_range(v) && in_range(w) {
+                        let v0 = f.vertices.0;
+                        let v1 = f.vertices.1;
+                        let v2 = f.vertices.2;
+                        let p = (self.vertices[v0].position * u)
+                            + (self.vertices[v1].position.to_vec() * v)
+                            + (self.vertices[v2].position.to_vec() * w);
+                        return Some(p);
+                    }
+                    face_index =
+                        next_face_classify(self, face_index, Some((u, v, w)), &mut checked);
+                }
+                _ => {
+                    face_index = next_face_classify(self, face_index, None, &mut checked);
+                }
+            }
+        }
+        None
+    }
+}
+
+#[inline]
+fn in_range<S>(v: S) -> bool
+where
+    S: BaseFloat,
+{
+    v >= S::zero() && v <= S::one()
+}
+
+#[inline]
+fn next_face_classify<S>(
+    polytope: &ConvexPolyhedron<S>,
+    face_index: usize,
+    bary_coords: Option<(S, S, S)>,
+    checked: &mut BitSet,
+) -> usize
```

Comment:
shouldn't it return `Option<usize>`?
overall, this function looks like it's trying to mimic an iterator. Perhaps, you should consider implementing it as such in the first place?

Reply:
Well, sort of yes. Just not sure how I would do that.

Reply:
I think it's not worth doing it as an Iterator, it's doing directed iteration, it takes the uvw as parameter, and building an iterator that does directed iteration is more work than the benefits it might give.